hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/Quota.java

```diff
@@ -50,7 +50,7 @@ public static Counts newInstance() {
    * The quota is violated if quota is set and usage &gt; quota.
    */
   public static boolean isViolated(final long quota, final long usage) {
-    return quota >= 0 && usage > quota;
+    return quota >= 0 && usage >= quota;
```

Comment:
The core of the change seems to be general HDFS not just RBF, right?

Reply:
Yes, the core changes are in general HDFS, but the caller to this method is in RBF and some method for WARNING logs. 

The general check of quota calls the method below.
`/**
   * Is quota violated?
   * The quota is violated if quota is set, delta &gt; 0 and
   * usage + delta &gt; quota.
   */
  static boolean isViolated(final long quota, final long usage,
      final long delta) {
    return quota >= 0 && delta > 0 && usage > quota - delta;
  }`

Reply:
I think this should be a general HDFS JIRA and then we could add tests for RBF later.

Reply:
ok, I'm new to this, I'm not sure what to do next，update the JIRA and PR,  then create another JIRA to add the UT?

Reply:
I think the method is correct only, It is used for LOGS in core HDFS parts, and Quota is said to be violated when the count is more than the value set, if the fileAndDirCount is 5 and quota is also 5, then quota isn't said to be violated.
The javadoc also tells that.

Regarding RBF, I think you need to change the call here to pass the delta, while verifying  :
https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterQuotaUsage.java#L99

Rather than the method which you tweaked use the below one with proper delta:
```
  static boolean isViolated(final long quota, final long usage,
      final long delta) {
    return quota >= 0 && delta > 0 && usage > quota - delta;
  }
```
That I think should solve the problem here. Not sure how things will work in case of storage quota though....

Wrote a UT in **TestRouterRPCMultipleDestinationMountTableResolver** , to check this, putting it up here as well in case anyone wants to try, might save some efforts.

```
  @Test
  public void testQuotaMultiDest() throws Exception {
    // NS-Quota as 4
    long nsQuota = 4;
    Path path = new Path("/test10");
    nnFs0.mkdirs(path);
    nnFs1.mkdirs(path);
    Map<String, String> destMap = new HashMap<>();
    destMap.put("ns0", "/test10");
    destMap.put("ns1", "/test10");
    MountTable addEntry = MountTable.newInstance("/test10", destMap);
    assertTrue(addMountTable(addEntry));
    RouterQuotaUpdateService updateService =
        routerContext.getRouter().getQuotaCacheUpdateService();
    updateService.periodicInvoke();

    // Set NS-Quota as 4
    RouterAdmin admin = getRouterAdmin();
    String[] argv = new String[] {"-setQuota", path.toString(), "-nsQuota",
        String.valueOf(nsQuota)};
    assertEquals(0, ToolRunner.run(admin, argv));
    updateService.periodicInvoke();
    resolver.loadCache(true);

    // Touch two files under ns-fed/tes10

    routerFs.create(new Path("/test10/file1")).close();
    routerFs.create(new Path("/test10/file2")).close();

    updateService.periodicInvoke();

    // This should have thrown exception since 2 target directories & 2 files: total 4 & Quota is 4
    routerFs.create(new Path("/test10/file3")).close();
  }
```

Reply:
Sorry for taking so long to update, I checked the code, router checks quota at getLocationsForPath, it is difficult to get  a delta. Here is the trace based on 3.3.0
```
org.apache.hadoop.hdfs.protocol.NSQuotaExceededException: The NameSpace quota (directories and files) is exceeded: quota=3 file count=4
	at org.apache.hadoop.hdfs.server.federation.router.RouterQuotaUsage.verifyNamespaceQuota(RouterQuotaUsage.java:100)
	at org.apache.hadoop.hdfs.server.federation.router.RouterRpcServer.getLocationsForPath(RouterRpcServer.java:1535)
	at org.apache.hadoop.hdfs.server.federation.router.RouterRpcServer.getLocationsForPath(RouterRpcServer.java:1475)
	at org.apache.hadoop.hdfs.server.federation.router.RouterClientProtocol.create(RouterClientProtocol.java:286)
	at org.apache.hadoop.hdfs.server.federation.router.RouterRpcServer.create(RouterRpcServer.java:611)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.create(ClientNamenodeProtocolServerSideTranslatorPB.java:494)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:532)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1070)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1020)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:948)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1845)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2952)
```